File: 8. String to Integer (atoi)/solution.c
/*
    Runtime: 4 ms, faster than 95.23% of C online submissions for String to Integer (atoi).
    Memory Usage: 7 MB, less than 66.38% of C online submissions for String to Integer 
*/

int myAtoi(char * ss){
	int i = 0, negative = 1;
    int64_t nbr = 0;
	while ((ss[i] == '\n') || (ss[i] == '\t') || (ss[i] == '\v') ||
				(ss[i] == ' ') || (ss[i] == '\f') || (ss[i] == '\r'))
			i++;
		if (ss[i] == '-')
			negative = -1;
		if (ss[i] == '+' || ss[i] == '-')
			i++;
		while (ss[i] && (ss[i] >= '0') && (ss[i] <= '9'))
		{
			int pop = (int)ss[i] - '0';
			if ((nbr > INT_MAX/10 || (nbr == INT_MAX / 10 && pop > 7)) && negative == 1) return INT_MAX;
			if ((nbr > INT_MAX/10 || (nbr == INT_MAX / 10 && pop > 8)) && negative == -1) return INT_MIN;
			nbr *= 10;
			nbr += pop;
			i++;
		}
		return nbr * negative;
}

File: 8. String to Integer (atoi)/solution.py
# Runtime: 36 ms, faster than 98.46% of Python3 online submissions for String to Integer (atoi).
# Memory Usage: 13.2 MB, less than 48.93% of Python3 online submissions for String to Integer (atoi).

class Solution:
    def myAtoi(self, s):
        s = s.lstrip()
        if len(s) == 0: return 0

        num = ''
        if s[0] == '-' or s[0] == '+':
            num = s[0]
            s = s[1:]

        for c in s:
            if c.isdigit():
                num += c
            else:
                break

        try:
            val = int(num)
            if (val.bit_length() >= 32):
                return (2**31 - 1) if val > 0 else -2**31
            return val
        except ValueError:
            return 0
        
